### ISSUE TITLE
Proposal: turn "const" at global level into a real storage class

### ISSUE THREAD

**Issue by kvark** [gpuweb/gpuweb#1394]
Here is the storage class description:

> Memory locations are partitioned into storage classes. Each storage class has unique properties determining mutability, visibility, the values it may contain, and how to use variables with it.

Let's say there is memory holding all the constants:
- [x] mutability (immutable)
- [x] visibility (visible to all)
- [x] values it may contain (only other constants or literals)
- [x] how to use variables with it (any way)

Is there a downside of then just assigning this type of memory a storage class? The upside is having more streamlined spec, less ways to declare a variable at the global scope. Note: this is only about global variables - for locals the `const` have a different meaning, and we've debated the need to separate them for a while.

 So with this proposal, instead of writing `const PI: f32 = 3.1415` we would do `var<const> PI: f32 = 3.1415`.

**Comment by dneto0** [comment]
The difference is that a current const definition at module scope is not associated with a memory location.
So I can't take its address, i.e. I can't form a pointer to it.

From the spec:

> A variable is a named reference to storage that can contain a value of a particular storable type.

So given:

      const PI: f32 = 3.1415;
      var<const>  AnotherPI: f32 = 3.1415;

       fn foo() {
            // would be possible, but we don't want this. Assumes #1368
            const p : ptr<const,f32> = AnotherPI;  
            // would be possible, but we don't want this. 
            // If you don't like #1368 and instead have explicit address-of.
            const p2 : ptr<const,f32> = &AnotherPI; 
       }

**Comment by kvark** [comment]
That's a great point!
Aren't we are already in a situation where for some variables taking a pointer is not possible? I'm talking about textures and samplers, in particular. Or do you mean that for all variables that have *explicit* storage class, there should be an ability to take a pointer from it?

**Comment by dneto0** [comment]
> Aren't we are already in a situation where for some variables taking a pointer is not possible? 

My plan of record is that you can always take a pointer of a variable. That's the consistent rule.  If there's storage, you can form a pointer to it.  That's what makes #1368 small and internally consistent.  (Ok, as small as it is now.  If we didn't have this rule then it would be bigger.)

Today, we can't put textures and samplers into arrays or structs. So if you squint there's no difference between the pointer and the value.  The value is an opaque "handle".  The difference between a pointer to handle and the handle itself is a matter of bookkeeping.

Where it needs exploration is: do we want to pass textures (and samplers) as parameters to helper functions.  And if we want reference semantics, do we encourage thinking of as getting passed as a pointer to texture/sampler to be dereferenced only when used by a builtin?  Or do we say the handle value itself already embodies the fact it's a reference to an opaque complex thing underneath.

Existing models:

- OpenCL : image types are only usable for function parameters (including kernel entry points).  It's a non-copyable handle that can only be passed into helper functions or image builtins. See [Restrictions](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_C.html#restrictions), point b.

- MSL:  texture types can be function arguments. Additionally, read-only storage images and sampled images can be the type for a variable in a function, but the variables must have an initializer.  (Latter restriction stated in an example only.)  
   - I tried an experiment and found that I could copy handles and select between them.
 
Here's an example MSL program that looks like it compiles to sensible AIR.  I haven't tried running it.
```C++
#include <metal_stdlib>

using T2D = metal::texture2d<float,metal::access::sample>;
using S = metal::sampler;

float4 helper(
     T2D im,
     S s) {
  return im.sample(s,float2(1.0));
}

fragment void foo(
     T2D im,
     T2D im2,
     S s,
     device float4 *A) {

  T2D ha = ((*A).x >0.0) ? im : im2;

  *A = helper(ha,s);
}
```


- SPIR-V / Vulkan (base functionality): 
     -  can pass textures and samplers to helper functions by pointer, or by value.   A texture or sampler *variable* (storage) is only in UniformConstant storage class, and only at module scope.
     - You can't declare a private (module-scope but thread private) or function storage class variable to hold a texture or sampler type. (You also can't store pointers.)
     - You can't select between pointers (e.g. that  ternary operator in the MSL example above would be invalid in SPIR-V.)
     - As a combined consequence of these and other rules, you can always trace an image value or sampler value inside a function back to the originating module-scope variable (pipeline-bound resource variable), or to a function parameter.  For a function parameter, you can always trace interprocedurally up through callees and then get to the module-scope resource variable at some level.  So in the end the usage model is spiritually very close to what OpenCL does:   the compiler can see through the trivial copies via OpCopyObject, for example.
     - Vulkan does permit arrays of resources, but WGSL does not.   For completeness though:  Without any variable-pointers extension, each pointer-type operand to a function call must either be: (the result ID of) a variable, a function parameter (pass through), or a pointer to an element of an array in UniformConstant (which covers both textures and samplers).

**Comment by kvark** [comment]
> My plan of record is that you can always take a pointer of a variable. That's the consistent rule. If there's storage, you can form a pointer to it.

The "handle" storage class is opaque, so "if there's storage" condition is quite non-trivial here.
And I agree with you that it comes down to whether or not we allow pointers of handles. If we can't get pointers to handles, then having "const" storage isn't going to be exceptional in any way, and it simplifies the number of different mechanics in the API.

So far, I'm not seeing any evidence on why we'd need to have pointers to handles, though. It appears to me that handles themselves are fine to pass around, just like in Metal.

**Comment by dneto0** [comment]
Another difference is lurking in how we define const_expr #1272

I think we need to make at least a module-scope const_expr for MVP, because that makes pipeline-overrideable constants useful.
For const_expr at module scope, you should not be able to reference a variable's contents (no loads).  But you do want to evaluate other const_exprs.
So current 'const' at module-scope ought to have const-expr initializers, that only refer to literals, certain operations, and the value of other already-defined module-scope 'const'.

A var is storage.  A const is a value (and not storage).   That separation makes a lot of other stuff clean.

**Comment by dneto0** [comment]
(For the allow-list of "certain operations", see OpSpecConstantOp in SPIR-V, the Shader subset.  )

**Comment by kvark** [comment]
> Another difference is lurking in how we define const_expr #1272

Correct me if that's wrong, but I thought that the current `const` can't be used in places where we need `const_expr` anyway. For example, the `const` can have specialization, which makes it a non-const-expr. So if we want to add a mechanism for const expressions somewhere down the road, we can do it, but it's not like `const` will have anything to do with them.

> A var is storage. A const is a value (and not storage). That separation makes a lot of other stuff clean.

Both a variable and a constant are data accessed by shaders, data that lives *somewhere*. If I declare a constant table of 32K, it will be placed in *some* memory for shader to read it (and you likely have a good idea where exactly it will be stored), so it will have storage. The separation from transparent storage (like a uniform buffer structure) and opaque storage ("handle") is already in place, so it may very well be possible that having `const` outside of that storage party isn't going to make anything cleaner.

**Comment by dneto0** [comment]
>  Correct me if that's wrong, but I thought that the current const can't be used in places where we need const_expr anyway.

Right, that's what's in the current grammar.  That should be fixed, which is what #1272 is about.  

> Both a variable and a constant are data accessed by shaders, data that lives somewhere.

"Accessed" when.  A 'const' at module scope is a value resolved at compile time (either create-shader-module time, or as late as pipeline-creation time for pipeline-overridable constants (a.k.a. specialization constants).

> If I declare a constant table of 32K vectors, it will be placed in some memory for shader to read it (and you likely have a good idea where exactly it will be stored), so it will have storage.

If the compiler can't tell which parts of it you're accessing, then that's correct. 
If you only ever access the first element and the compiler can prove it, then the compiler has the opportunity to only "store" the one value. And because that's so small, and will contain a value that the compiler knows at the earliest possible moment 
(immediately after parsing), then it will fold that into downstream computations, and there will be no storage.

I can take the address of a variable. I can't take the address of a 'const'.

**Comment by kvark** [comment]
Ok, it sounds like all the arguments for positions are voiced. We are just blocked on #1272 and #1368 to proceed.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-vf2f-2022-02-08,09-minutes#burndown-of-httpsgithubcomorgsgpuwebprojects2">WGSL meeting minutes 2022-02-08</a></summary>



    * DN: No, because we don’t have Storage for const
    * DM: At a global level we don’t have storage so much as named expressions, so we can close this.
</details>
